### PR TITLE
Implement teardown for gh-ost library

### LINKED ISSUE #465: Fixing leaked goroutines
Hey all!

Thanks for building such an awesome library!

We're getting ready to start using this as a part of our production workflow, and we wanted to build a wrapper around the `gh-ost` utility to manage some configuration on our end and to integrate it into our development/deployment workflows. We were hoping to build this wrapper in Go and consume the library directly.

When building out the proof of concept, we noticed that `migrator.go` and a few other places are starting infinite goroutines (for example the `initiateStatus`) that seem to be relying on the fact that it is invoked from a CLI and that the program will exit when `Migrate()` is complete.

The wrapper we are building will be a long running process (http server) that will manage the invocations of the gh-ost library. Would you be willing to accept a PR that cleans up these infinite goroutines, allowing the core logic to be used outside of the single CLI run?

Happy to hear any feedback on this approach if not. Thanks again!

### PR DESCRIPTION
**Pull request by nikhilmat**
Fixes #465 

This PR adds explicit cleanup to gh-ost for when the library is consumed by another application and might be executed multiple times (for example, by a webserver). This stops any infinite goroutines that have started once the migration has finished, closes any DB connections it has opened, and ensures that the migration context that is passed around is not being globally accessed - instead it instantiates a new one for each migration run and passes it around.

Please let me know how this looks to you, and happy to make any changes to improve it!

It also pulls in the change from https://github.com/outbrain/golib/pull/4.

### CODE REVIEW

**Comment by shlomi-noach** [comment]
Thank you! just as heads up it'll take me a few days to review this, sorry for the delay.

**Comment by nikhilmat** [comment]
@shlomi-noach do you think you will get a chance to review this this week?

**Comment by shlomi-noach** [comment]
this is the week!

File: vendor/github.com/outbrain/golib/sqlutils/sqlutils.go [hunk 1]
```diff
@@ -144,6 +144,17 @@ func GetDB(mysql_uri string) (*sql.DB, bool, error) {
 	return knownDBs[mysql_uri], exists, nil
 }
 
+// Resets the knownDBs cache, used when the DB connections have been closed,
+//   and new connections are needed to access the DB
+func ResetDBCache() {
+	knownDBsMutex.Lock()
+	defer func() {
+		knownDBsMutex.Unlock()
+	}()
```
Comment (shlomi-noach) [inline vendor/github.com/outbrain/golib/sqlutils/sqlutils.go]: simpler:
```
defer knownDBsMutex.Unlock()
```

File: go/logic/migrator.go [hunk 2]
```diff
@@ -1211,3 +1230,24 @@ func (this *Migrator) finalCleanup() error {
...
+		log.Infof("Tearing down applier")
+		this.applier.Teardown()
+	}
+
+	if this.eventsStreamer != nil {
+		log.Infof("Tearing down streamer")
+		this.eventsStreamer.Teardown()
+	}
+
+	sqlutils.ResetDBCache()
```
Comment (shlomi-noach) [inline go/logic/migrator.go]: :+1:

File: go/logic/applier.go [hunk 3]
```diff
@@ -30,16 +30,18 @@ const (
...
 type Applier struct {
-	connectionConfig *mysql.ConnectionConfig
-	db               *gosql.DB
-	singletonDB      *gosql.DB
-	migrationContext *base.MigrationContext
+	connectionConfig  *mysql.ConnectionConfig
+	db                *gosql.DB
+	singletonDB       *gosql.DB
+	migrationContext  *base.MigrationContext
+	finishedMigrating bool
```
Comment (shlomi-noach) [inline go/logic/applier.go]: This variable is accesses read/write by different goroutines. You'd need o change it to `int64` and use `atomic.StoreInt64` and `atomic.LoadInt64` ; see various such examples throughout the code.

File: go/logic/applier.go [hunk 4]
```diff
@@ -310,6 +312,9 @@ func (this *Applier) InitiateHeartbeat() {
 
 	heartbeatTick := time.Tick(time.Duration(this.migrationContext.HeartbeatIntervalMilliseconds) * time.Millisecond)
 	for range heartbeatTick {
+		if this.finishedMigrating {
```
Comment (shlomi-noach) [inline go/logic/applier.go]: unsafe operation ; concurrency

File: go/logic/applier.go [hunk 5]
```diff
@@ -1039,3 +1044,10 @@ func (this *Applier) ApplyDMLEventQueries(dmlEvents [](*binlog.BinlogDMLEvent))
 	log.Debugf("ApplyDMLEventQueries() applied %d events in one transaction", len(dmlEvents))
 	return nil
 }
+
+func (this *Applier) Teardown() {
+	log.Debugf("Tearing down...")
+	this.db.Close()
+	this.singletonDB.Close()
+	this.finishedMigrating = true
```
Comment (shlomi-noach) [inline go/logic/applier.go]: unsafe operation ; concurrency ; use `atomic.StoreInt64`

File: go/logic/migrator.go [hunk 6]
```diff
@@ -83,11 +84,13 @@ type Migrator struct {
 	applyEventsQueue chan *applyEventStruct
 
 	handledChangelogStates map[string]bool
+
+	finishedMigrating bool
```
Comment (shlomi-noach) [inline go/logic/migrator.go]: see concurrency concerns above

File: go/logic/migrator.go [hunk 7]
```diff
@@ -718,6 +727,9 @@ func (this *Migrator) initiateStatus() error {
 	this.printStatus(ForcePrintStatusAndHintRule)
 	statusTick := time.Tick(1 * time.Second)
 	for range statusTick {
+		if this.finishedMigrating {
```
Comment (shlomi-noach) [inline go/logic/migrator.go]: unsafe operation ; concurrency

File: go/logic/migrator.go [hunk 8]
```diff
@@ -942,6 +954,9 @@ func (this *Migrator) initiateStreaming() error {
 	go func() {
 		ticker := time.Tick(1 * time.Second)
 		for range ticker {
+			if this.finishedMigrating {
```
Comment (shlomi-noach) [inline go/logic/migrator.go]: unsafe operation ; concurrency

File: go/logic/migrator.go [hunk 9]
```diff
@@ -1132,6 +1147,10 @@ func (this *Migrator) executeWriteFuncs() error {
 		return nil
 	}
 	for {
+		if this.finishedMigrating {
```
Comment (shlomi-noach) [inline go/logic/migrator.go]: unsafe operation ; concurrency

File: go/logic/migrator.go [hunk 10]
```diff
@@ -1211,3 +1230,24 @@ func (this *Migrator) finalCleanup() error {
 
 	return nil
 }
+
+func (this *Migrator) teardown() {
+	this.finishedMigrating = true
```
Comment (shlomi-noach) [inline go/logic/migrator.go]: unsafe operation ; concurrency

File: vendor/github.com/outbrain/golib/sqlutils/sqlutils.go (hunk 1 above)
Comment (nikhilmat) [reply vendor/github.com/outbrain/golib/sqlutils/sqlutils.go]: of course, thanks!

**Comment by nikhilmat** [comment]
@shlomi-noach thanks for the notes on the concurrency concerns - those have been fixed!

i've been doing some additional testing over the past week and came across an issue with the DB cache in `sqlutils` when running concurrent migrations. since we are now closing our DB connections at the end of each migration, if another migration is running when one ends, the connection gets pulled out from underneath them because the cache has them share a connection pool. 

Do you have any objections to attaching the db connections pools that connect to `migrationContext.DatabaseName` and the `information_schema` to the migration context, and removing the database cache in `sqlutils`? that way, we avoid creating too many pools by always referencing it on the context, and it will be scoped to each migration so they don't interfere if one migration finishes and closes its pool.

**Comment by nikhilmat** [comment]
@shlomi-noach, I implemented that proposed changed here https://github.com/github/gh-ost/pull/479/commits/84bdfdb1ada1669bb27458aaf9c0750b6f276de6 - let me know what you think of the approach! i believe with these changes, each migration will maintain its own connections before cleaning them up at the end - allowing for migrations to run concurrently.

Thanks again for your review, and sorry for the delay with the above changes! Will be sure to get to any feedback you have shortly :)

**Comment by shlomi-noach** [comment]
@nikhilmat I'll be in a conference this week so will only take a look next week. But meanwhile, Re: thread pool, please see https://github.com/github/gh-ost/pull/491 and https://github.com/github/gh-ost/issues/482, work in progress by @zmoazeni to address that.

**Comment by zmoazeni** [comment]
@shlomi-noach @nikhilmat hey folks. I'm not sure how the thread pool changes affect this PR, but by all means if you need any help just ask.

As far as I understand the codebase and my threadpool PR, I don't explicitly create a new connection. It just reuses the existing ones. So I _think_ they will both go hand-in-hand. But if they don't, just let me know.

**Comment by shlomi-noach** [comment]
@nikhilmat 

> Do you have any objections to attaching the db connections pools that connect to migrationContext.DatabaseName and the information_schema to the migration context, and removing the database cache in sqlutils? 
>  I implemented that proposed changed here 84bdfdb - let me know what you think of the approach! i believe with these changes, each migration will maintain its own connections before cleaning them up at the end - allowing for migrations to run concurrently.

Thank you for working on this. 
I do not have objections to this, and can see the reasoning.

One thing I'd like not to change is `sqlutils.go`, which, as you notice, is under `vendor`. I'd like to keep it as it is (though I confess I've tweaked it here and there in other projects, and it's a technical debt I have to consolidate all changes).

Also, whatever happened to the `finishedMigrating` flag? The code looks so different now that I feel this is a completely new review.

**Comment by nikhilmat** [comment]
@shlomi-noach great, I believe all the changes on moving the DB connections are covered in this PR - I was able to successfully run concurrent migrations locally.

Totally understand about not wanting to change the source code in `vendor/`. Are you okay with submitting this change to https://github.com/outbrain/golib and then updating the dependencies here? If not, I can stop using `sqlutils.GetDB` altogether, and just rely on that method exposed on the `migrator`. Let me know which you prefer, and I'll make the change. (Relatedly, I had a question about some of the vendor dependencies here: https://github.com/github/gh-ost/issues/501)

Sorry for the confusion with the layered changes. The `finishedMigrating` flag is still present on the `applier` and `migrator`. I updated them to use `atomic. LoadInt64` and `atomic. StoreInt64` in this commit https://github.com/github/gh-ost/pull/479/commits/5788ab534726f40a555db04861862c28d0ecf125. Besides addressing the concurrency concerns, there were no changes in the last two commits since your previous review (unless I missed something).

Thanks again! Let me know how you'd like to proceed with the `sqlutils` changes.

**Comment by nikhilmat** [comment]
@shlomi-noach any thoughts on the above with regards to the `vendor/` changes for the DB cache? Let me know if there's anything else here that needs revising :)

**Comment by nikhilmat** [comment]
@shlomi-noach friendly ping :) anything else i can do to help this along?

**Comment by shlomi-noach** [comment]
@nikhilmat sorry for the delay! I was out for a while and then busy catching up. Will get to it.

**Comment by nikhilmat** [comment]
@shlomi-noach no problem, thanks!

**Comment by shlomi-noach** [comment]
Again, apologies for the delay. Still on it, just have some serious backlog.

**Comment by nikhilmat** [comment]
@shlomi-noach thanks for checking in on this. i know there's still a decision to be made on how to apply the changes to the vendor directory:

> Totally understand about not wanting to change the source code in vendor/. Are you okay with submitting this change to https://github.com/outbrain/golib and then updating the dependencies here? If not, I can stop using sqlutils.GetDB altogether, and just rely on that method exposed on the migrator. Let me know which you prefer, and I'll make the change.

Feel free to chime in on your preference on this and I can make the change before taking another pass. Otherwise, I can go ahead remove the usage of `sqlutils.GetDB` and just rely on the migration context specific db cache.

Thanks!

**Comment by shlomi-noach** [comment]
Apologies for keeping you waiting on such a small issue. 
For now, I'd like you to restore `sqlutils.go` to its original state.

What I also see you've done is https://github.com/github/gh-ost/pull/479/files#diff-b9f3c8a49a606c2654d3c9c33727e879R253, which just duplicates the code in `sqlutils.go`? But also undermines your desire to avoid caching the DB object.

Furthermore, `MigrationContext` isn't the right place for generating the DB instance ; consider it more as a data object that can be serialized.

I suggest the following:
- Restore `sqlutils.go` to its original state
- Remove the `GetDB()` code from `MigrationContext`
- Create `GetDB()` in `go/mysql/utils.go`
- Remove reference to `sqlutils.GetDB` and refactor as `mysql.GetDB`

Thank you!

**Comment by nikhilmat** [comment]
@shlomi-noach the goal with moving the caching to the `MigrationContext` was that the caching would be done a per migration level, and so if migrations were running concurrently, closing a connection in one migration would not affect another. However, happy to follow your suggestion and revert the sqlutils changes and move that method to the mysql utils.

Thanks!

**Comment by shlomi-noach** [comment]
> the goal with moving the caching to the MigrationContext was that the caching would be done a per migration level, and so if migrations were running concurrently, closing a connection in one migration would not affect another

I understand better now, thank you for explaining. To elaborate abit, a futuristic idea is to have the `MigrationContext` instance serialized and persisted. To that effect, it would not be possible to assume database driver/connections would exist when the instance loads. The specific use case right now is https://github.com/github/gh-ost/issues/302.

**Comment by nikhilmat** [comment]
Ah, cool! Thanks for sharing, that's good to know. I had a separate question about adding a custom, migration-specific logger (we're sending our logs to cloudwatch), and this is really helpful. Will open up an issue to discuss.

This is ready for another pass when you get a chance. Incorporated your feedback and updated with the latest master!

**Comment by nikhilmat** [comment]
👋 @shlomi-noach - do you think you'll get a chance to take another look at this soon?

**Comment by shlomi-noach** [comment]
@nikhilmat apologies yet again for the delays and I appreciate your patience, thank you for pinging.
This looks good to me!

What I'll do is merge this onto a local `nm-refactor-migration-context` branch, and send to testing.
You will see this PR as "Merged", but it won't be merged to master. I'll @mention you on the merged branch.
This complexity is required because of the decoupling we make between this public repo and our internal infrastructure.

**Comment by nikhilmat** [comment]
wonderful - thank you!!!

**Pull request merged**